test/model/message_test.dart

```diff
@@ -76,6 +94,314 @@ void main() {
+    ]);
+
+    store.dispose();
+    check(async.pendingTimers).single.duration.equals(const Duration(seconds: 1));
+  }));
+
+  group('sendMessage', () {
+    final stream = eg.stream();
+    final streamDestination = StreamDestination(stream.streamId, eg.t('some topic'));
+    late StreamMessage message;
+
+    test('outbox messages get unique localMessageId', () async {
+      await prepare(stream: stream);
+      await prepareMessages([]);
+
+      for (int i = 0; i < 10; i++) {
+        connection.prepare(json: SendMessageResult(id: 1).toJson());
+        await store.sendMessage(destination: streamDestination, content: 'content');
+      }
+      // [store.outboxMessages] has the same number of keys (localMessageId)
+      // as the number of sent messages, which are guaranteed to be distinct.
+      check(store.outboxMessages).keys.length.equals(10);
+    });
+
+    void checkState({required OutboxMessageState expectedState}) {
+      check(store.outboxMessages).values.single.state.equals(expectedState);
+    }
+
+    Future<void> prepareOutboxMessage({
+      MessageDestination? destination,
+      int? zulipFeatureLevel,
+    }) async {
+      message = eg.streamMessage(stream: stream);
+      await prepare(stream: stream, zulipFeatureLevel: zulipFeatureLevel);
+      await prepareMessages([eg.streamMessage(stream: stream)]);
+      connection.prepare(json: SendMessageResult(id: 1).toJson());
+      await store.sendMessage(
+        destination: destination ?? streamDestination, content: 'content');
+    }
+
+    late Future<void> outboxMessageFailFuture;
+    Future<void> prepareOutboxMessageToFailAfterDelay(Duration delay) async {
+      message = eg.streamMessage(stream: stream);
+      await prepare(stream: stream);
+      await prepareMessages([eg.streamMessage(stream: stream)]);
+      connection.prepare(apiException: eg.apiBadRequest(), delay: delay);
+      outboxMessageFailFuture = store.sendMessage(
+        destination: streamDestination, content: 'content');
+    }
+
+    test('smoke DM: hidden -> waiting -> (delete)', () => awaitFakeAsync((async) async {
+      await prepareOutboxMessage(destination: DmDestination(
+        userIds: [eg.selfUser.userId, eg.otherUser.userId]));
+      checkState(expectedState: OutboxMessageState.hidden);
+
+      async.elapse(kLocalEchoDebounceDuration);
+      checkState(expectedState: OutboxMessageState.waiting);
+
+      await store.handleEvent(eg.messageEvent(
+        eg.dmMessage(from: eg.selfUser, to: [eg.otherUser]),
+        localMessageId: store.outboxMessages.keys.single));
+      check(store.outboxMessages).isEmpty();
+    }));
+
+    test('smoke stream message: hidden -> waiting -> (delete)', () => awaitFakeAsync((async) async {
+      await prepareOutboxMessage();
+      checkState(expectedState: OutboxMessageState.hidden);
+
+      async.elapse(kLocalEchoDebounceDuration);
+      checkState(expectedState: OutboxMessageState.waiting);
```

Comment:
I like this structure where the helpers center on a single outbox-message. It helps keep the tests focused, since the bulk of them are all about a single outbox-message (because there's very little between-message interaction to be tested).